pkg/auth/globarolebinding.go

```diff
@@ -0,0 +1,119 @@
+package auth
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	rancherv3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/webhook/pkg/authentication"
+	mgmtcontrollers "github.com/rancher/webhook/pkg/generated/controllers/management.cattle.io"
+	v3 "github.com/rancher/webhook/pkg/generated/controllers/management.cattle.io/v3"
+	"github.com/rancher/wrangler-api/pkg/generated/controllers/rbac"
+	"github.com/rancher/wrangler/pkg/webhook"
+	admissionv1 "k8s.io/api/admission/v1"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/authentication/user"
+	k8srequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/client-go/rest"
+	"k8s.io/kubernetes/pkg/registry/rbac/validation"
+	rbacregistryvalidation "k8s.io/kubernetes/pkg/registry/rbac/validation"
+	"k8s.io/utils/trace"
+)
+
+func NewGRBValidator(cfg *rest.Config) (webhook.Handler, error) {
+	grb, err := mgmtcontrollers.NewFactoryFromConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	rbac, err := rbac.NewFactoryFromConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	rbacRestGetter := authentication.RBACRestGetter{
+		Interface: rbac.Rbac(),
+	}
+
+	ruleResolver := rbacregistryvalidation.NewDefaultRuleResolver(rbacRestGetter, rbacRestGetter, rbacRestGetter, rbacRestGetter)
+
+	return &globalRoleBindingValidator{
+		globalRoleClient: grb.Management().V3().GlobalRole(),
+		ruleSolver:       ruleResolver,
+	}, nil
+
+}
+
+type globalRoleBindingValidator struct {
+	globalRoleClient v3.GlobalRoleClient
+	ruleSolver       validation.AuthorizationRuleResolver
+}
+
+func (grbv *globalRoleBindingValidator) Admit(response *webhook.Response, request *webhook.Request) error {
+	listTrace := trace.New("globalRoleBindingValidator Admit", trace.Field{Key: "user", Value: request.UserInfo.Username})
+	defer listTrace.LogIfLong(1 * time.Second)
+
+	newGRB, err := grbObject(request)
+	if err != nil {
+		return err
+	}
+
+	// Pull the global role to get the rules
+	globalRole, err := grbv.globalRoleClient.Get(newGRB.GlobalRoleName, metav1.GetOptions{})
```

Comment:
Do we need to use a cache? You know better than me how many times this might be called.  Are we creating a lot of grbs?

Reply:
Per offline discussion, this only applies around user creation and wouldn't affect cluster/project creation so leaving it as a live call